### PR TITLE
solution

### PR DESCRIPTION
**Pull request by savchuksavchuk**
[DEMO LINK](https://savchuksavchuk.github.io/react_add-todo-form/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,61 +1,131 @@
...
+  if (allowedCharacters.includes(char.toLocaleLowerCase())) {
+    return true;
+  }
+
+  return false;
+}
 
 export const App = () => {
+  const [toDos, addToDo] = useState([...todosFromServer]);
+  const [toDoTitle, setTitle] = useState('');
```
Comment (liliia-livandovskaya) [inline src/App.tsx:21]:
```suggestion
  const [title, setTitle] = useState('');
```

File: src/App.tsx [hunk 2]
```diff
@@ -1,61 +1,131 @@
...
+  const [toDoUser, setUser] = useState('');
```
Comment (liliia-livandovskaya) [inline src/App.tsx:22]:
```suggestion
  const [user, setUser] = useState('');
```

File: src/App.tsx [hunk 3]
```diff
@@ -1,61 +1,131 @@
...
+              placeholder="Please enter a title"
+              value={toDoTitle}
+              onChange={(event) => {
+                const character = event.target.value.slice(-1);
+
+                if (isCharacterAllowed(character)) {
+                  setTitle(event.target.value);
+                  setTitleValidation(1);
+                }
+              }}
```
Comment (liliia-livandovskaya) [inline src/App.tsx:48]: this looks weak, instead, create a handler, and the same way all over your code

File: src/components/UserInfo/UserInfo.tsx [hunk 4]
```diff
@@ -1 +1,18 @@
...
+  email: string,
+};
+
+type Props = {
+  user: User | undefined,
+};
+
+export const UserInfo: React.FC<Props> = ({ user }) => {
+  return (
+    <a className="UserInfo" href={`mailto:${user !== undefined ? user.email : ''}`}>
```
Comment (liliia-livandovskaya) [inline src/components/UserInfo/UserInfo.tsx:14]: add formatting

**Review by liliia-livandovskaya — CHANGES_REQUESTED** [review]
GJ! Pay attention to the comments

File: src/App.tsx [hunk 5]
```diff
@@ -1,61 +1,131 @@
...
 import './App.scss';
 
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+import { TodoList } from './components/TodoList';
+
+const allowedCharacters
+  = 'abcdefghijkmnopqrstuvwxyzабвгґдеєжзиіїйклмнопрстуфхцчшщьюя123456789 ';
```
Comment (polosanya) [inline src/App.tsx:9]: Do you really need to hardcore all the possible chars?
Try to find another solution

File: src/App.tsx [hunk 6]
```diff
@@ -1,61 +1,131 @@
...
+function isCharacterAllowed(char: string) {
+  if (allowedCharacters.includes(char.toLocaleLowerCase())) {
+    return true;
+  }
+
+  return false;
+}
 
 export const App = () => {
+  const [toDos, addToDo] = useState([...todosFromServer]);
```
Comment (polosanya) [inline src/App.tsx:20]:
```suggestion
  const [todos, setTodos] = useState([...todosFromServer]);
```
Same below. Function name must be set according to the noun variable.
Please fix all cases

File: src/App.tsx [hunk 7]
```diff
@@ -1,61 +1,131 @@
...
+        >
+          <label>
+            Title:
+            <input
+              type="text"
+              data-cy="titleInput"
+              placeholder="Please enter a title"
+              value={toDoTitle}
+              onChange={(event) => {
+                const character = event.target.value.slice(-1);
```
Comment (polosanya) [inline src/App.tsx:42]: Create separate functions to handle events

File: src/App.tsx [hunk 8]
```diff
@@ -1,61 +1,131 @@
...
+
+                if (isCharacterAllowed(character)) {
+                  setTitle(event.target.value);
+                  setTitleValidation(1);
+                }
+              }}
+            />
+          </label>
+
+          {isTitleValid === -1 && (
```
Comment (polosanya) [inline src/App.tsx:52]: What does -1 mean? Why can't you use boolean value?

File: src/App.tsx [hunk 9]
```diff
@@ -1,61 +1,131 @@
...
-            <option value="0" disabled>Choose a user</option>
-          </select>
-
-          <span className="error">Please choose a user</span>
+          <label>
+            User:
+            <select
+              data-cy="userSelect"
+              value={toDoUser}
+              onChange={(event) => {
```
Comment (polosanya) [inline src/App.tsx:63]: Separate function.
Please fix all cases

**Review by polosanya — CHANGES_REQUESTED** [review]
Tests failed. Before making a Pull Request (or a new commit to an already existing PR), you need to write the `npm test` command in the terminal and make sure that all tests pass. After that, the mentor will be able to review the homework. If you have questions please ask them in fe_chat

**Comment by savchuksavchuk** [comment]
Hello, i can't manage with two last tests. I don't understand how they are running. Should i join one of the Q&A with this question?

**Comment by TarasHoliuk** [comment]
> Hello, i can't manage with two last tests. I don't understand how they are running. Should i join one of the Q&A with this question?

It looks like it should work. I don't see the reason for another behavior. The code is good, so I'll approve.

**Review by TarasHoliuk — APPROVED** [review]

**Pull request closed without merging**